system-parachains/coretime/coretime-kusama/src/coretime.rs

```diff
@@ -0,0 +1,252 @@
+		// `ref_time` = 10177115 + (1 * 25000000) + (2 * 100000000) + (80 * 13932) = 236291675
+		// `proof_size` = 3612
+		// Add 5% to each component and round to 2 significant figures.
+		let call_weight = Weight::from_parts(248_000_000, 3800);
+
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: call_weight,
+				call: assign_core_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, Location::parent(), message) {
+			Ok(_) => log::debug!(
+				target: "runtime::coretime",
+				"Core assignment sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Core assignment failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn check_notify_revenue_info() -> Option<(RCBlockNumberOf<Self>, Self::Balance)> {
+		let revenue = CoretimeRevenue::get();
+		CoretimeRevenue::set(&None);
+		revenue
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn ensure_notify_revenue_info(when: RCBlockNumberOf<Self>, revenue: Self::Balance) {
+		CoretimeRevenue::set(&Some((when, revenue)));
+	}
+}
+
+/// Implements the [`AdaptPrice`] trait to control the price changes of bulk coretime. This tweaks
+/// the [`pallet_broker::Linear`] implementation which hard-corrects to 0 if cores are offered but
+/// not sold for just one sale. The monotonic lead-in is increased in magnitude to enable faster
+/// price finding. The change in base price between sales has a lower limit of 0.5 to allow downward
+/// pressure to be applied, while keeping a conservative upper limit of 1.2 (movements capped at 20%
+/// if cores sell out) to avoid runaway prices in the early sales. The intention is that this will
+/// be coupled with a low number of cores per sale and a 100% ideal bulk ratio for the first sales.
+pub struct PriceAdapter;
+impl AdaptPrice for PriceAdapter {
+	fn leadin_factor_at(when: FixedU64) -> FixedU64 {
+		// Start at 5x the base price and decrease to the base price at the end of leadin.
+		FixedU64::from(5).saturating_sub(FixedU64::from(4) * when)
+	}
+
+	fn adapt_price(sold: CoreIndex, target: CoreIndex, limit: CoreIndex) -> FixedU64 {
+		if sold <= target {
+			// Range of [0.5, 1.0].
+			FixedU64::from_rational(1, 2).saturating_add(FixedU64::from_rational(
+				(sold).into(),
+				(target.saturating_mul(2)).into(),
+			))
+		} else {
+			// Range of (1.0, 1.2].
+
+			// Unchecked math: In this branch we know that sold > target. The limit must be >= sold
+			// by construction, and thus target must be < limit.
+			FixedU64::one().saturating_add(FixedU64::from_rational(
+				(sold - target).into(),
+				(limit - target).saturating_mul(5).into(),
+			))
+		}
+	}
+}
+
+impl pallet_broker::Config for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	type Currency = Balances;
+	type OnRevenue = CreditToCollatorPot;
```

Comment:
We had a RFC for burning the fees. Questionable if we should/want to enable this without a public referendum. Maybe we should do this and highlight in the upgrade that this will burn the fees. Then the community could reject it if not wanted.

Reply:
Another question, don't we need to register the coretime chain as a location that supports teleporting the native token? Or is this automatically done by it being a system chain? CC @joepetrowski @bkontur

Reply:
It should be a teleporter, it wouldn't make a difference in burning though AFAICT.

But Donal already has a branch that burns it, we should just merge that into this (see https://github.com/polkadot-fellows/runtimes/pull/212#issuecomment-2011307981)

Reply:
The branch doesn't actually burn the funds yet, just sets it aside in a holding account.
This way it can be burnt by a future version, so we'd have a runtime upgrade which is more directly about burning.

Reply:
Ah right, but we don't want to merge this with crediting revenue to the collator pot. Holding account "to be burned" is OK IMO.

Reply:
> Another question, don't we need to register the coretime chain as a location that supports teleporting the native token? Or is this automatically done by it being a system chain? CC @joepetrowski @bkontur


coretime is correctly configured to trust other system parachains + relaychain for KSM teleports (`type IsTeleporter = ConcreteAssetFromSystem<KsmRelayLocation>;`), and also other system parachains trust coretime for KSM teleports (they also use `ConcreteAssetFromSystem`) and relaychain is also [configured](https://github.com/polkadot-fellows/runtimes/blob/main/relay/kusama/src/xcm_config.rs#L148) to trust coretime with KSM teleports

Reply:
Holding account merged in [e373dec](https://github.com/polkadot-fellows/runtimes/commit/1a1b2ea9bd3efc21fb517b5cd22b053c8e373dec).